pymeasure/instruments/hcp/tc038d.py

```diff
@@ -54,90 +52,100 @@ class TC038D(Instrument):
-        if isinstance(values, int):
-            data += [0x0, self.byteMode // 2]  # 2B number of elements
-            data.append(self.byteMode)  # 1B number of write data
-            for i in range(self.byteMode - 1, -1, -1):
-                data.append(values >> i * 8 & 0xFF)
-        elif isinstance(values, collections.abc.Sequence):
+    def write(self, command):
+        """Write a command to the device.
+
+        :param str command: comma separated string of:
+            - the function: read ('R') or write ('W') or 'echo',
+            - the address to write to (e.g. '0x106' or '262'),
+            - the values (comma separated) to write
+            - or the number of elements to read (defaults to 1).
+        """
+        function, address, *values = command.split(",")
+        data = [self.address]  # 1B device address
+        data.append(self.functions[function])  # 1B function code
+        address = int(address, 16) if "x" in address else int(address)
+        data.extend(address.to_bytes(2, "big"))  # 2B register address
+        if function.upper() == "W":
             elements = len(values) * self.byteMode // 2
             data += [elements >> 8, elements & 0xFF]  # 2B number of elements
-            data.append(len(values) * self.byteMode)  # 1B number of write data
+            data.append(elements * 2)  # 1B number of bytes to write
             for element in values:
-                for i in range(self.byteMode - 1, -1, -1):
-                    data.append(element >> i * 8 & 0xFF)
-        else:
-            raise TypeError(("Values has to be an integer or an iterable of "
-                             f"integers. values: {values}"))
+                data.extend(int(element).to_bytes(self.byteMode, "big", signed=True))
+        elif function.upper() == "R":
+            count = int(values[0]) * self.byteMode // 2 if values else self.byteMode // 2
+            data.extend(count.to_bytes(2, "big"))  # 2B number of elements to read
+        elif function == "echo":
+            data[-2:] = [0, 0]
+            if values:
+                data.extend(int(values[0]).to_bytes(2, "big"))  # 2B test data
         data += CRC16(data)
         self.write_bytes(bytes(data))
+
+    def read(self):
+        """Read response and interpret the number"""
+        # Slave address, function
         got = self.read_bytes(2)
-        # slave address, function
-        if got[1] == self.functions['writeMultiple']:
+        if got[1] == self.functions['R']:
+            # length of data to follow
+            length = self.read_bytes(1)
+            # data length, 2 Byte CRC
+            read = self.read_bytes(length[0] + 2)
+            if read[-2:] != bytes(CRC16(got + length + read[:-2])):
+                raise ConnectionError("Response CRC does not match.")
+            return int.from_bytes(read[:-2], byteorder="big", signed=True)
+        elif got[1] == self.functions['W']:
```

Comment:
So, this part is the "ACK" response of this instrument to sent commands? If yes, might be good to call that out in a comment.

Reply:
Yes, if got[1] is the writing function code, the read message is the acknowledgement.